src/persist/src/s3.rs

```diff
@@ -163,30 +163,216 @@ impl S3BlobCore {
+            // the code works perfectly well if we just pretend this was a
+            // multipart upload of 1 part.
+            1
+        } else {
+            first_part.parts_count()
         };
+        tracing::trace!(
+            "s3 download first header took {:?} ({} parts)",
+            start_overall.elapsed(),
+            num_parts
+        );
+        if num_parts < 1 {
+            return Err(format!("unexpected number of s3 object parts: {}", num_parts).into());
+        }
 
-        let val = object
-            .body
-            .collect()
-            .await
-            .map_err(|e| format!("collecting byte buffer: {}", e))?
-            // TODO: `into_bytes().to_vec()` results in one extra copy than
-            // necessary. Changing `Blob::get` to return `Bytes` directly would
-            // be more efficient, or adding a `into_vec()` method upstream.
-            .into_bytes()
-            .to_vec();
+        // TODO: Plumb an Arc<AsyncRuntime> on S3BlobCore instead. I tried but
+        // was getting a "Cannot drop a runtime in a context where blocking is
+        // not allowed" error that I didn't want to get distracted with. All the
+        // calls into the s3 library require that this context is set anyway, so
+        // I suppose this is no worse than where we started.
+        let async_runtime = AsyncHandle::try_current().map_err(|err| err.to_string())?;
+
+        // Continue to fetch the first part's body while we fetch the other
+        // parts.
+        let start_part_body = Instant::now();
+        let mut part_bodies = Vec::new();
+        let first_body_len = usize::try_from(first_part.content_length)
+            .map_err(|err| format!("unexpected s3 content_length: {}", err))?;
+        let mut total_body_len = first_body_len;
+        part_bodies.push(
+            async_runtime.spawn_named(
+                || "persist_s3blob_get_body",
+                first_part
+                    .body
+                    .collect()
+                    .map(move |res| (start_part_body.elapsed(), first_body_len, res)),
+            ),
+        );
+
+        if num_parts > 1 {
+            // Fetch the headers of the rest of the parts. (Starting at part 2
+            // because we already did part 1.)
+            let start_headers = Instant::now();
+            let mut part_futs = Vec::new();
+            for part_num in 2..=num_parts {
+                let part_fut = async_runtime.spawn_named(
+                    || "persist_s3blob_get_header",
+                    client
+                        .get_object()
+                        .bucket(&self.bucket)
+                        .key(&path)
+                        .part_number(part_num)
+                        .send()
+                        .map(move |res| (start_headers.elapsed(), res)),
+                );
+                part_futs.push(part_fut);
+            }
+
+            // As each part header comes back, first off a task to fetch the
```

Comment:
okay looked into doing the later TODO. I was hoping that we could switch Blob::get to return a bytes::Buf impl instead of Vec<u8>, but it turns out parquet decoding needs io::Seek in addition to io::Read, so it's harder than I was hoping

then I looked into doing this suggestion. the docs for GetObject indicate that it doesn't return the total object size in any way if you ask for just a part. so, we could do a HeadObject request to get it up front, but that would probably be a pessimization for small requests. maybe we do the HeadObject request once we discover that there are multiple parts? I think that would work but since the real answer here is avoiding copying them all into one place, I'm not sure it's worth the interim complexity

summary: I'm leaning toward adding this idea to the TODO but not doing it yet

Reply:
:ok_hand: thanks for looking into it!